In [15]:
import tensorflow as tf
from tensorflow.examples.tutorials.mnist import input_data
#load data
mnist = input_data.read_data_sets("MNIST_data", one_hot=True)

batch_size = 100

batch_count = mnist.train.num_examples // batch_size

#define place holder
x = tf.placeholder(tf.float32, [None, 784])
y = tf.placeholder(tf.float32, [None, 10])

#create a 1-level neural network
W = tf.Variable(tf.truncated_normal([784, 10], stddev=0.1))
# b = tf.Variable(tf.zeros([10]))
b = tf.Variable(tf.zeros([10])+0.1)

wx_b = tf.matmul(x, W)+b

#激活函数，实现非线性化
# prediction = tf.nn.relu(wx_b)
# prediction = tf.nn.tanh(wx_b)
# prediction = tf.nn.sigmoid(wx_b)
prediction = tf.nn.softmax(wx_b)

loss = tf.reduce_mean(tf.nn.softmax_cross_entropy_with_logits_v2(labels=y, logits=prediction))
# loss = tf.reduce_mean(tf.nn.sigmoid_cross_entropy_with_logits(labels=y, logits=prediction))

#train
train = tf.train.GradientDescentOptimizer(0.2).minimize(loss)

init = tf.global_variables_initializer()

#比较预测值与实际值
correct_pred = tf.equal(tf.argmax(y, 1), tf.argmax(prediction, 1)) #argmax 返回沿着某个维度最大值的位置

#准确率
accuracy = tf.reduce_mean(tf.cast(correct_pred, tf.float32))

with tf.Session() as sess:
    sess.run(init)
    
    for step in range(20):
        for batch in range(batch_count):
            batch_xs, batch_ys = mnist.train.next_batch(batch_size)
            sess.run(train, feed_dict={x:batch_xs, y:batch_ys})
        
        acc = sess.run(accuracy, feed_dict={x:mnist.test.images, y:mnist.test.labels})
        print("step:" + str(step) + " accurancy:" + str(acc))



Extracting MNIST_data\train-images-idx3-ubyte.gz
Extracting MNIST_data\train-labels-idx1-ubyte.gz
Extracting MNIST_data\t10k-images-idx3-ubyte.gz
Extracting MNIST_data\t10k-labels-idx1-ubyte.gz
step:0 accurancy:0.7325
step:1 accurancy:0.8128
step:2 accurancy:0.8873
step:3 accurancy:0.8996
step:4 accurancy:0.9051
step:5 accurancy:0.9079
step:6 accurancy:0.9107
step:7 accurancy:0.9129
step:8 accurancy:0.9148
step:9 accurancy:0.9152
step:10 accurancy:0.9164
step:11 accurancy:0.9175
step:12 accurancy:0.918
step:13 accurancy:0.9179
step:14 accurancy:0.9176
step:15 accurancy:0.9184
step:16 accurancy:0.92
step:17 accurancy:0.9198
step:18 accurancy:0.92
step:19 accurancy:0.9203


sigmoid:
step:0 accurancy:0.7355
step:1 accurancy:0.8283
step:2 accurancy:0.8504
step:3 accurancy:0.8633
step:4 accurancy:0.8696
step:5 accurancy:0.8761
step:6 accurancy:0.8789
step:7 accurancy:0.882
step:8 accurancy:0.8848
step:9 accurancy:0.8864
step:10 accurancy:0.8891
step:11 accurancy:0.8904
step:12 accurancy:0.8925
step:13 accurancy:0.893
step:14 accurancy:0.8942
step:15 accurancy:0.896
step:16 accurancy:0.8963
step:17 accurancy:0.8975
step:18 accurancy:0.8988
step:19 accurancy:0.899

step:0 accurancy:0.744
step:1 accurancy:0.8259
step:2 accurancy:0.8472
step:3 accurancy:0.8601
step:4 accurancy:0.8664
step:5 accurancy:0.8718
step:6 accurancy:0.8766
step:7 accurancy:0.8797
step:8 accurancy:0.8824
step:9 accurancy:0.8855
step:10 accurancy:0.887
step:11 accurancy:0.889
step:12 accurancy:0.8907
step:13 accurancy:0.8918
step:14 accurancy:0.8938
step:15 accurancy:0.8939
step:16 accurancy:0.8948
step:17 accurancy:0.8958
step:18 accurancy:0.8967
step:19 accurancy:0.8972

tanh:
step:0 accurancy:0.7856
step:1 accurancy:0.8157
step:2 accurancy:0.8336
step:3 accurancy:0.8315
step:4 accurancy:0.8461
step:5 accurancy:0.8396
step:6 accurancy:0.8424
step:7 accurancy:0.8394
step:8 accurancy:0.8384
step:9 accurancy:0.8427
step:10 accurancy:0.8441
step:11 accurancy:0.845
step:12 accurancy:0.8412
step:13 accurancy:0.8394
step:14 accurancy:0.835
step:15 accurancy:0.8442
step:16 accurancy:0.8515
step:17 accurancy:0.8362
step:18 accurancy:0.8475
step:19 accurancy:0.8463

step:0 accurancy:0.7821
step:1 accurancy:0.802
step:2 accurancy:0.8245
step:3 accurancy:0.828
step:4 accurancy:0.8486
step:5 accurancy:0.844
step:6 accurancy:0.8423
step:7 accurancy:0.85
step:8 accurancy:0.8476
step:9 accurancy:0.841
step:10 accurancy:0.8428
step:11 accurancy:0.846
step:12 accurancy:0.8471
step:13 accurancy:0.8462
step:14 accurancy:0.8382
step:15 accurancy:0.8403
step:16 accurancy:0.8402
step:17 accurancy:0.8444
step:18 accurancy:0.8416
step:19 accurancy:0.8506

relu:
step:0 accurancy:0.6199
step:1 accurancy:0.68
step:2 accurancy:0.7026
step:3 accurancy:0.7106
step:4 accurancy:0.7147
step:5 accurancy:0.7145
step:6 accurancy:0.7183
step:7 accurancy:0.7224
step:8 accurancy:0.7231
step:9 accurancy:0.7225
step:10 accurancy:0.7253
step:11 accurancy:0.7235
step:12 accurancy:0.725
step:13 accurancy:0.7278
step:14 accurancy:0.7276
step:15 accurancy:0.7294
step:16 accurancy:0.7271
step:17 accurancy:0.7257
step:18 accurancy:0.7285
step:19 accurancy:0.7292

step:0 accurancy:0.5043
step:1 accurancy:0.6182
step:2 accurancy:0.6314
step:3 accurancy:0.6352
step:4 accurancy:0.7112
step:5 accurancy:0.7908
step:6 accurancy:0.8017
step:7 accurancy:0.8042
step:8 accurancy:0.8078
step:9 accurancy:0.8088
step:10 accurancy:0.8122
step:11 accurancy:0.8125
step:12 accurancy:0.8139
step:13 accurancy:0.8125
step:14 accurancy:0.8151
step:15 accurancy:0.8154
step:16 accurancy:0.8129
step:17 accurancy:0.8153
step:18 accurancy:0.8166
step:19 accurancy:0.8185